disorder-jack/src/Disorder/Jack/Combinators.hs

```diff
@@ -0,0 +1,157 @@
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# LANGUAGE LambdaCase #-}
+module Disorder.Jack.Combinators (
+    arbitrary
+
+  , sized
+  , sizedInt
+  , sizedIntegral
+  , sizedNat
+  , sizedNatural
+
+  , chooseChar
+  , chooseInt
+  , chooseIntegral
+
+  , oneof
+  , elements
+  , list
+  , list1
+  , vector
+  ) where
+
+import           Control.Monad (replicateM)
+import           Control.Applicative (Applicative(..))
+
+import           Data.Bool (not)
+import           Data.Char (Char, ord, chr)
+import           Data.Function (($), (.))
+import           Data.Functor (Functor(..))
+import           Data.Int (Int)
+import qualified Data.List as List
+import           Data.List.NonEmpty (NonEmpty(..))
+import qualified Data.List.NonEmpty as NonEmpty
+
+import           Disorder.Jack.Core
+import           Disorder.Jack.Shrink
+import           Disorder.Jack.Tree
+
+import           Prelude (Num(..), Integral, min, max)
+import qualified Prelude as Savage
+
+import           System.Random (Random)
+
+import qualified Test.QuickCheck as QC
+
+
+-- | Construct a 'Jack' using a type's QuickCheck 'QC.Arbitrary' instance.
+arbitrary :: QC.Arbitrary a => Jack a
+arbitrary =
+  mkJack QC.shrink QC.arbitrary
+
+-- | Construct a 'Jack' that depends on the size parameter.
+sized :: (Int -> Jack a) -> Jack a
+sized f =
+  Jack $ QC.sized (runJack . f)
+
+-- | Generates an 'Int'. The number can be positive or negative and its maximum
+--   absolute value depends on the size parameter.
+sizedInt :: Jack Int
+sizedInt =
+  sizedIntegral
+
+-- | Generates an integral number. The number can be positive or negative and
+--   its maximum absolute value depends on the size parameter.
+sizedIntegral :: Integral a => Jack a
+sizedIntegral =
+  mkJack QC.shrinkIntegral QC.arbitrarySizedIntegral
+
+-- | Generates a non-negative 'Int'. The number's maximum value depends on the
+--   size parameter.
+sizedNat :: Jack Int
+sizedNat =
+  sizedNatural
+
+-- | Generates a natural number. The number's maximum value depends on
+--   the size parameter.
+sizedNatural :: Integral a => Jack a
+sizedNatural =
+  mkJack QC.shrinkIntegral QC.arbitrarySizedNatural
+
+-- | Generates a 'Char' in the given range.
+chooseChar :: Char -> Char -> Jack Char
+chooseChar b0 b1 =
+  fmap chr $ chooseIntegral (ord b0) (ord b1)
+
+-- | Generates an 'Int' in the given range.
+chooseInt :: Int -> Int -> Jack Int
```

Comment:
Yeah unfortunately we have stricter requirements to be able to shrink sensibly, `chooseIntegral` is the closest thing, maybe that should just be called `choose` would work 99% of the time